src/hostname/hostnamed.c

```diff
@@ -250,8 +250,27 @@ static int get_firmware_vendor(char **ret) {
          return get_hardware_firmware_data("bios_vendor", ret);
 }
 
-static int get_firmware_date(char **ret) {
-         return get_hardware_firmware_data("bios_date", ret);
+static int get_firmware_date(uint64_t *ret) {
+         const char *k;
+         struct tm tm;
+         time_t v;
+         _cleanup_free_ char *bios_date = NULL;
+         int r;
+
+         r = get_hardware_firmware_data("bios_date", &bios_date);
+         if (r <= 0)
+                 return r;
+
+         k = strptime(bios_date, "%m/%d/%Y", &tm);
+         if (!k || *k)
+                 return r;
```

Comment:
honestly, I'd just write a manual parser, i.e.:

```c
const char *p = bios_date;
r = extract_many_words(&p, "/", EXTRACT_DONT_COALESCE_SEPARATORS, &month, &day, &year);
if (r < 0)
        return r;
if (r != 3) /* less than three args read? */
        return -EINVAL;
if (!isempty(p)) /* more left in the string? */
        return -EINVAL;
 
unsigned m, d, y;
r = safe_atou(month, &m);
if (r < 0) 
        return r;
if (m < 1 || m > 12)
        return -EINVAL;

r = safe_atou(day, &d);
if (r < 0)
        return r;
if (d < 1 || d > 31)
        return -EINVAL;

r = safe_atou(year, &y);
if (r < 0)
        return r;
if (y < 1970 || y > (usigned) INT_MAX)
        return -EINVAL;  

tm = (struct tm) {
        .tm_mday = d,
        .tm_mon = m,
        .tm_year = y,
};
v = timegm(&tm);
if (v == (time_t) -1))
        return -errno;
if (tm.tm_mday != (int) d || tm.tm_mon != (int m) || tm.tm_year != (int) y)
        return -EINVAL; /* date was not normalized? (e.g. "30th of feb") */

Reply:
parse_timestamp() can parse a lot of stuff. I think we should be more restrictive here, and just parse exactly what we can expect.

parse_timestamp() will happily parse timezones and µs and whatnot, none of which we should accept here I think.

this data is ultimately from a low-quality source: the firmware, hence I think we should be careful here, and better suppress garbage data than to parse it incorrectly.

Reply:
I do not have strong opinion over that. But if you provide a restrictive parser, then please add tests for it.